evalml/automl/automl_search.py

```diff
@@ -522,9 +527,46 @@ def search(self, data_checks="auto", show_iteration_plot=True):
 
         best_pipeline = self.rankings.iloc[0]
         best_pipeline_name = best_pipeline["pipeline_name"]
+        self._best_pipeline = self.get_pipeline(best_pipeline['id'])
+        if self._train_best_pipeline:
+            X_train = _convert_to_woodwork_structure(X)
+            y_train = _convert_to_woodwork_structure(y)
+            X_threshold_tuning = None
+            y_threshold_tuning = None
+            if self.optimize_thresholds and self.objective.is_defined_for_problem_type(ProblemTypes.BINARY) and self.objective.can_optimize_threshold:
+                X_train, X_threshold_tuning, y_train, y_threshold_tuning = train_test_split(X_train, y_train, test_size=0.2, random_state=self.random_state)
+            self._best_pipeline.fit(X_train, y_train)
+            self._best_pipeline = self._threshold_pipeline(self._best_pipeline, X_threshold_tuning, y_threshold_tuning)
         logger.info(f"Best pipeline: {best_pipeline_name}")
         logger.info(f"Best pipeline {self.objective.name}: {best_pipeline['score']:3f}")
 
+    def _threshold_pipeline(self, pipeline, X_threshold_tuning, y_threshold_tuning, fold=None):
+        """Tunes the threshold of a binary pipeline to the X and y thresholding data
+
+        Arguments:
+            pipeline (Pipeline): Pipeline instance to threshold
+            X_threshold_tuning (ww DataTable): X data to tune pipeline to
+            y_threshold_tuning (ww DataColumn): Target data to tune pipeline to
+            fold (None, int): The fold of training this belongs to in AutoMLSearch. Defaults to None
+
+        Returns:
+            Trained pipeline instance
+        """
+        if self.objective.is_defined_for_problem_type(ProblemTypes.BINARY):
+            pipeline.threshold = 0.5
+            if X_threshold_tuning:
+                if fold is not None:
+                    logger.debug(f"\t\t\tFold {fold}: Optimizing threshold for {self.objective.name}")
+                y_predict_proba = pipeline.predict_proba(X_threshold_tuning)
+                if isinstance(y_predict_proba, pd.DataFrame):
+                    y_predict_proba = y_predict_proba.iloc[:, 1]
+                else:
+                    y_predict_proba = y_predict_proba[:, 1]
+                pipeline.threshold = self.objective.optimize_threshold(y_predict_proba, y_threshold_tuning, X=X_threshold_tuning)
+                if fold is not None:
+                    logger.debug(f"\t\t\tFold {fold}: Optimal threshold found ({pipeline.threshold:.3f})")
```

Comment:
This is minor, but I suggest we move this log to `_compute_cv_scores` and delete the other one
```python
cv_pipeline = self._threshold_pipeline(cv_pipeline, X_threshold_tuning, y_threshold_tuning, fold=i)
if pipeline.threshold is not None and self.optimize_thresholds:
    logger.debug(f"\t\t\tFold {i}: Optimal threshold found ({pipeline.threshold:.3f})")
```

That way the logging code is all in `_compute_cv_scores` and this helper just has impl code.